lib/rules/no-shadow.js

```diff
@@ -54,9 +72,85 @@ module.exports = {
         const options = {
             builtinGlobals: context.options[0] && context.options[0].builtinGlobals,
             hoist: (context.options[0] && context.options[0].hoist) || "functions",
-            allow: (context.options[0] && context.options[0].allow) || []
+            allow: (context.options[0] && context.options[0].allow) || [],
+            ignoreOnInitialization: context.options[0] && context.options[0].ignoreOnInitialization
         };
 
+        /**
+         * Checks whether or not the nodes types include the node type.
+         * @param {ASTNode} node a node to get.
+         * @param {[string]} types an array of nodes types.
+         * @returns {boolean} True if nodes types include the node type.
+         */
+        function isNodeMatchesTypes(node, types) {
+            return types.includes(node.type);
+        }
+
+        /**
+         * Searches from the current node through its ancestry to find a matching node.
+         * @param {ASTNode} node a node to get.
+         * @param {MatchCallback} match a callback that checks whether or not the node verifies its condition or not.
+         * @returns {ASTNode} the matching node.
+         */
+        function findSelfOrAncestor(node, match) {
+            let currentNode = node;
+
+            while (currentNode && !match(currentNode)) {
+                currentNode = currentNode.parent;
+            }
+            return currentNode;
+        }
+
+        /**
+         * Checks if a variable and a shadowedVariable have the same init pattern ancestor.
+         * @param {Object} variable a variable to check.
+         * @param {Object} shadowedVariable a shadowedVariable to check.
+         * @returns {boolean} Whether or not the variable and the shadowedVariable have the same init pattern ancestor.
+         */
+        function isInitPatternNode(variable, shadowedVariable) {
+            const outerDef = shadowedVariable.defs[0];
+            const innerDef = variable.defs[0];
+            let fun = innerDef && innerDef.node;
+
+            while (fun && !isNodeMatchesTypes(fun, FUN_NODES_TYPES)) {
+                if (isNodeMatchesTypes(fun, FUN_DEC_NODES_TYPES)) {
+                    fun = null;
+                } else {
+                    fun = fun.parent;
+                }
+            }
+
+            if (fun) {
+                const { parent } = fun;
+
+                const nestedFun = findSelfOrAncestor(parent, node =>
+                    isNodeMatchesTypes(node, FUN_NODES_TYPES));
+
+                if (!nestedFun) {
```

Comment:
This could find an arrow function or function expression that is not nested between definitions, but surrounds them and therefore should not affect the logic of this option.

```js
/* eslint no-shadow: ["error", { ignoreOnInitialization: true }] */

const x = foo(x => x); // allowed, correct

() => {
    const y = foo(y => y); // not allowed, incorrect?
}

```